[3단계 - 리팩터링] 띠용(주용은) 미션 제출합니다

jye326 opened the pull request:
대규모 공사 해봤습니다! 잘 부탁드립니다~

1. `RegisterController`와 `LoginController`를 통해 회원가입 및 로그인 기능 처리
2. `AbstractController` 추가하여 GET, POST 메서드 처리 분리
3. `StaticRenderer`를 통한 정적 페이지 렌더링 및 리다이렉션 처리
4. `Http11Processor`에서 로직 단순화 및 컨트롤러 사용
5. `HttpStatusCode`, `Request` 리팩토링 및 `Response` 확장
6. 쿼리 파라미터 파싱 유틸 클래스 `StringParser` 추가

jye326 committed:
feat: 회원가입 및 로그인 로직 분리

1. `RegisterController`와 `LoginController`를 통해 회원가입 및 로그인 기능 처리
2. `AbstractController` 추가하여 GET, POST 메서드 처리 분리
3. `StaticRenderer`를 통한 정적 페이지 렌더링 및 리다이렉션 처리
4. `Http11Processor`에서 로직 단순화 및 컨트롤러 사용
5. `HttpStatusCode`, `Request` 리팩토링 및 `Response` 확장
6. 쿼리 파라미터 파싱 유틸 클래스 `StringParser` 추가

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/AbstractController.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public void service(Request request, Response response) throws Exception {
+        String method = request.getHttpMethod();
+        if ("GET".equalsIgnoreCase(method)) {
```
GET, POST enum 으로 관리해도 좋을 것 같아요~~

jye326 replied:
반영했습니다!

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/LoginController.java:
```diff
@@ -0,0 +1,89 @@
+
+        // 쿼리 파라미터 로그인
+        String uri = request.getResourcePath();
+        if (uri.contains("?")) {
```
`POST /login` 으로 요청보내면 query 관련 로직이 없어도 되지 않을까요?

jye326 replied:
쿼리파라미터 로그인은 1단계때의 로그인 요구사항(GET http://localhost:8080/login?account=gugu&password=password )을 그대로 지원하고 싶어서 삭제하지 않았습니다!

POST 요청은 doPost로 다른 메서드가 처리하기 때문에 query 로직을 타지 않습니다!

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/LoginController.java:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+    private boolean login(Map<String, String> params, Response response) {
```
controller 말고 다른 계층이 갖고 있으면 좋은 로직같은데, 어떻게 생각하시나요?

jye326 replied:
아직은 계층 분리가 필요하지 않다고 생각하여 유지했습니다! 추후 복잡도가 올라가면 분리될 것 같습니다!

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/LoginController.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Override
+    protected void doPost(Request request, Response response) throws Exception {
+        if (login(parseQueryParameter(request.getBody()), response)) {
```
로그인에 실패하면 어떤 응답을 받을 수 있나요?

jye326 replied:
응답코드를 401로 수정 후 로그인 페이지를 렌더링하도록 수정했습니다!

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/RegisterController.java:
```diff
@@ -0,0 +1,33 @@
+    @Override
+    protected void doPost(Request request, Response response) throws Exception {
+        if (register(parseQueryParameter(request.getBody()))) {
+            staticRenderer.redirectToIndexPage(response);
```
Early Return 하면 좋을 것 같아요~ 흐름상 아래까지 무조건 내려가는 것 같아요

jye326 replied:
헉! 완전 놓쳤네요! 감사합니다~

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/RegisterController.java:
```diff
@@ -0,0 +1,33 @@
+        throw new IllegalArgumentException(ErrorMessage.INVALID_REGISTER_REQUEST.getMessage());
+    }
+
+    private boolean register(Map<String, String> params) {
```
해당 부분도 다른 계층이 관리하면 좋을 로직같은데, 어떻게 생각하시는지 궁금합니다

jye326 replied:
마찬가지로 지금 필수는 아니라고 생각합니다~

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/StaticRenderer.java:
```diff
@@ -0,0 +1,55 @@
+        if (contentType == null) {
+            contentType = "text/html";
+        }
+        return contentType + ";charset=utf-8";
```
contentTyp 으로 `charset=utf-8` 붙일지 말지 판단할 수 있을 것 같아요~

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/http11/Response.java:
```diff
@@ -16,8 +22,9 @@ public class Response {
 
-    public Response() {
+    public Response(OutputStream outputStream) {
         headers = new LinkedHashMap<>();
```
LinkedHashMap 으로 관리하는 이유가 있을까요?

jye326 replied:
응답 Header의 출력 순서를 관리하기 위함입니다!

minjae8563 replied:
저는 저희 java 코드가 출력 순서를 고려하지 않아도 괜찮다고 생각해서요

물론 띠용의 의견도 공감합니다 👍🏻

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/http11/Response.java:
```diff
@@ -72,4 +79,29 @@ public String getContentLength() {
+                        .stream()
+                        .map(entry -> entry.getKey() + ": " + entry.getValue() + " ")
+                        .collect(Collectors.joining("\r\n")));
+        if (getCookieMapSize() != 0) {
```
headers 주석에서는 cookie 별도 관리라 하셔서, cookie 만 있는 map 인 걸로 이해가 가요.
근데 entrySet() 을 순회할 이유가 있을까요?

jye326 replied:
CookieManager 내부의 쿠키 맵에는 여러 개의 쿠키를 Key: Val 형태로 저장할 수 있습니다. 위 코드 부분에서 현재까지 저장된 모든 쿠키를 SetCookie 포맷으로 출력합니다!

minjae8563 reviewed (approved):
좋은 구조네요 💯 재밌게 읽었습니다.
저도 코드 구현중이라 Pull 없이 리뷰한 터라, 잘못 이해한 부분이 있을 수 있습니다😭

반영하실 부분 반영하면 좋을 것 같습니다~!

jye326 committed:
feat: HttpMethod enum 추가 및 관련 로직 리팩토링

1. HTTP 메서드 관리를 위한 `HttpMethod` enum 클래스 추가
2. 기존 문자열 기반 메서드 처리를 `HttpMethod` enum을 사용하도록 변경
3. `Request`와 `AbstractController`에서 HTTP 메서드 처리 로직 개선 및 가독성 향상

jye326 committed:
feat: HTTP 상태 코드 추가 및 로직 개선

1. `HttpStatusCode`에 401 Unauthorized 상태 코드 추가
2. `LoginController` 및 `RegisterController`의 로직에서 상태 코드 설정 처리 추가
3. `StaticRenderer`의 리다이렉션 및 페이지 렌더링 방식 개선
4. `Http11Processor`의 요청 처리 로직 리팩토링 및 상태 코드 설정 처리 추가
5. 사용하지 않는 코드 및 불필요한 import 제거
